### ISSUE TITLE
Equivalent of React's PureRenderMixin

### ISSUE THREAD

**Issue by japgolly** [japgolly/scalajs-react#9]
> If your React component's render function is "pure" (in other words, it renders the same result given the same props and state), you can use this mixin for a performance boost in some cases.

http://facebook.github.io/react/docs/pure-render-mixin.html

**Comment by japgolly** [comment]
Bloody callbacks stuff up Props equality... Might need to introduce a mechanism for callback identity and comparison.

**Comment by Odomontois** [comment]
+1 First thing i was sad about.

**Comment by japgolly** [comment]
A better equivalent of this will be in the next version of scalajs-react. (Better because the JS version 1) has an issue, 2) doesn't play with with Scala, 3) give no guarantees).

Release should be out in under a month.

**Comment by ngbinh** [comment]
Awesome!

**Comment by Odomontois** [comment]
Right now i'm using this workaround.
```scala
implicit class ReactCompBExt[Props, State, Backend](comp: ReactComponentB[Props, State, Backend]) {    
    def pure = shouldUpdateS { case (was, toBe) => was == toBe }

    def shouldUpdate(f: ((Props, Props)) => Boolean) = comp.shouldComponentUpdate(
      (scope, newProps, _) => f(scope.props, newProps))

    def shouldUpdateS(f: (((Props, State), (Props, State))) => Boolean) = comp.shouldComponentUpdate(
      (scope, newProps, newState) => f((scope.props, scope.state), (newProps, newState))
    )
  }
```
I could say nothing about performance yet though

**Comment by japgolly** [comment]
Using universal equality is dangerous in that when you, maybe now, maybe later, slip something in your props or state for which `==` doesn't hold, you're not going to know unless you pay close attention at  runtime and then investigate back. The Scala compiler is not your friend and will not help you.

This is especially an issue when you pass callbacks around (which is the norm). How would one expect Scala to compare functions? By reference? Most callbacks will be regenerated unless you've structured them a very special way, and again, because the compiler's not involved there's *no* way of knowing whether you're doing it wrong and are going to break `shouldComponentUpdate` on every call.

It's unpleasant but it's a real issue. It's something to be aware of when using an approach like above. 

I have a solution for these issues. I'll try to get it online well before the release so they can be critiqued. Feedback is always appreciated and just now looking at your snippet there's already a case I didn't touch - state comparison (I only use props comparison).

**Comment by japgolly** [comment]
I've created a branch for the performance management stuff that will go into 0.9.0 and added tools that address this issue.

Read all about it here:
https://github.com/japgolly/scalajs-react/blob/perf/extra/PERF.md

This will go out in 0.9.0 in about a month's time, and close this issue.

**Comment by japgolly** [comment]
Closed via 9e8f5bd85c70fe85f5538e11065f9e5841d27d4c
